pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -196,6 +202,16 @@ def restore_loops(self) -> None:
         self.trainer.fit_loop.global_step = self._loaded_checkpoint["global_step"]
         self.trainer.fit_loop.current_epoch = self._loaded_checkpoint["epoch"]
 
+        assert self.trainer.state.fn is not None
```

Comment:
nit: let's raise an explicit exception instead of relying on assert. ideally assert is only used inside of tests

Reply:
this is just to avoid mypy issues. https://github.com/PyTorchLightning/pytorch-lightning/pull/9413#discussion_r711102946